### PR TITLE
[Music]Store song bitrate, samplerate and channels in library for filtering

### PR DESCRIPTION
**Pull request by the-black-eagle**
## Description
This adds the ability for Kodi to read the bitrate, sample rate and No. of channels from music files when scanning them into the library and stores the values in the music database.

## Motivation and Context
The only way at present to see these values is by actually playing a song and looking at the values returned by the music player.  This is less than ideal if a user wishes, for example, to find all their multi-channel songs or high sample rate audio.  The only workaround currently is to add something to the filenames but this is laborious and easy to get wrong or overlook.  With the additions in this PR, users will now be able to use smart playlists to group their songs by sample rate, bitrate or the number of channels.  It's also possible to make nodes to do the same thing so you can have nodes for high sample rate audio (96 Khz or above), quadraphonic albums, 5.1 albums etc etc.  

## How Has This Been Tested?
Tested locally on Ubuntu 18.04 with both SQLite and MySQL databases.  Also tested against music db upgrades (v72 -> v75) with and without re-scanning the music files.  Created nodes and smart playlists to list albums where the files sample rate is > 44.1 Khz or the files are multi-channel audio.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

File: xbmc/guilib/guiinfo/MusicGUIInfo.cpp [hunk 1]
```diff
@@ -313,6 +313,37 @@ bool CMusicGUIInfo::GetLabel(std::string& value, const CFileItem *item, int cont
...
+        }
+        break;
+      }
+      case LISTITEM_SAMPLERATE:
+      {
+        int sampleRate = tag->GetSampleRate();
+        if (sampleRate > 0)
+        {
+          value = StringUtils::Format("%.5g", static_cast<double>(sampleRate) / 1000.0);
+
```
Comment (DaveTBlake) [inline xbmc/guilib/guiinfo/MusicGUIInfo.cpp]: Whitespace

File: xbmc/interfaces/json-rpc/schema/version.txt [hunk 2]
```diff
@@ -1 +1 @@
-JSONRPC_VERSION 11.4.0
+JSONRPC_VERSION 11.5.0
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/schema/version.txt]: You cross-posted with the merge of 11.5.0, so 11.6.0 here

**Review by DaveTBlake — COMMENTED** [review]
Good stuff @the-black-eagle 
No rush with these minors, I may as well do some final testing first.

But did give the PR a better title - Kodi (in the player) has always read these values, the "feature" is saving them in the db and being able to make smartplaylists and custom nodes based on them

File: xbmc/guilib/guiinfo/MusicGUIInfo.cpp (hunk 1 above)
Comment (the-black-eagle) [reply xbmc/guilib/guiinfo/MusicGUIInfo.cpp]: Sorry.  Fixed.

File: xbmc/interfaces/json-rpc/schema/version.txt (hunk 2 above)
Comment (the-black-eagle) [reply xbmc/interfaces/json-rpc/schema/version.txt]: Done.

**Comment by scott967** [comment]
Would it be possible to return sample rate as 44.1, 48, 96 in the infolabel?  I think this is more common way of expressing it and it takes up less screen real estate.   Since AFAIK sample rates are pretty much fixed it could just be a lookup.

**Comment by DaveTBlake** [comment]
@scott967 sample rate is reported for listitem in the same format as it has been for the current playing song (Musicplayer.samplerate info label), so for the purposes of this PR I think it is right as it is.

However both places replaced could have the format changed in a subsequent PR - I know @the-black-eagle wanted a different format for the listitem and I discouraged him as consistency is more important, and also I had assumed that skinners had some choice over how they formated stuff. It could be interesting to have icons for set values (if they can be enumerated) rather than numbers.

There are mods to Estuary to follow that make use of the new info labels this PR provides, that would be an opportunity to look at nice ways to present this data.

**Comment by the-black-eagle** [comment]
@scott967 That is already how the listitem is returned for sample rate, 44.1/48/96/192 etc. Bit rate is similarly returned in kbps eg 320, 192 etc.  If the file is CBR encoded then the actual bitrate is returned, if it's VBR then the average bitrate over the whole file is returned. Channels is just a straight int eg 1, 2, 4, 6. As Dave said, this is consistent with how the runtime musicplayer values are returned.
@DaveTBlake Yes, it's possible to show flags instead of the actual values.  Estuary does that for the musicplayer values already.  My somewhat simple mod that I will PR if this is merged just shows the actual values returned in the song info dialog and the bottom left of the nav pane when highlighting a song.

File: xbmc/GUIInfoManager.cpp [hunk 3]
```diff
@@ -5871,6 +5871,28 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
...
+///                  \anchor ListItem_BitRate
+///                  _string_,
+///     @return The bitrate of a song. Actual rate for CBR, average rate for VBR.
+///     <p><hr>
+///     @skinning_v19 **[New Infolabel]** \link ListItem_BitRate `ListItem.BitRate`\endlink
+///     <p>
+///   }///   \table_row3{   <b>`ListItem.SampleRate`</b>,
+///                  \anchor ListItem_SampleRate
+///                  _string_,
+///     @return The sample rate of a song eg 44100, 48000, 96000 etc.
```
Comment (DaveTBlake) [inline xbmc/GUIInfoManager.cpp]: This probably needs a adjustment since what the info label returns is 44.1 etc. 
Sorry for such a tedious comment

**Review by DaveTBlake — APPROVED** [review]
A minor comment change, then good to go. Thanks for this @the-black-eagle

File: xbmc/GUIInfoManager.cpp (hunk 3 above)
Comment (the-black-eagle) [reply xbmc/GUIInfoManager.cpp]: No need to be sorry, you're quite correct.  Comment updated :)

**Comment by scott967** [comment]
Same as MusicPlayer is perfect.  Thanks.

**Comment by Ch1llb0** [comment]
> Channels is just a straight int eg 1, 2, 4, 6. As Dave said, this is consistent with how the runtime musicplayer values are returned.

Asking myself whether this could be changed entirely? Have been wondering why the whole skin engine always only returns a channel count (e.g. 5 channels) which could be a 5.0 layout or a 4.1 layout. Both would be valid although rare, but with the current way it's implemented now for the infolabels, too, there's no way to tell which layout it is. It's the same for basically all layouts that may carry an LFE channel. Especially music does have some very odd layouts sometimes...

**Comment by Ch1llb0** [comment]
Wondering whether retrieving and saving the audio codec to the library via an addition to this might be possible? The video library has that, but the music library is still missing this information.

**Comment by the-black-eagle** [comment]
@Ch1llb0 It's unlikely that I'll add it for several reasons.

- You can use a rule such as `<rule field="filename" operator="endswith">  <value>flac</value> </rule>` to divide your music up by flac/mp3/whatever already
- It would take much longer to scan the files as the tagreader doesn't have codec info available so each file would have to be opened twice, once to read the tags and once to determine the codec
- By using a combination of the rule above and sample rate/ bit rate or both, you'll probably know what the codec is anyway, eg 96k flac, 320k mp3 etc
- There would be a lot of unnecessary replication in the db. If you have an album or set of any size then it's more than likely just in the one format so storing the codec for each track (when they are all the same) seems a bit pointless however some users may have ripped some tracks to mp3 and then later finished by ripping to flac.  As long as the tags are correct, Kodi will group it as one album so each track would _need_ the codec to be stored just in case.

Because of all that, I think it's impractical to store the codec in the music library in a similar way to the video library.  Sorry.

**Comment by Ch1llb0** [comment]
Ok, understood. I do understand the technical argument although it's still a bummer to not have the codec information in the same fashion it's available for videos. I'll not add any "guessed" information, but only what's specifically offered by the skin engine. User may however guess from those information...

Thank you for your quick reply!

**Pull request merged**